// @sanity/form-builder/lib/components/ActivateOnFocus/ActivateOnFocus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ActivateOnFocus;

var _react = _interopRequireDefault(require("react"));

var _ActivateOnFocus = require("./ActivateOnFocus.styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// This is transitional in order to track usage of the ActivateOnFocusPart part from within the form-builder package
function ActivateOnFocus(props) {
  var children = props.children,
      message = props.message,
      onActivate = props.onActivate,
      isOverlayActive = props.isOverlayActive;

  function handleClick() {
    if (onActivate) {
      onActivate();
    }
  }

  function handleBlur() {
    if (onActivate && isOverlayActive) {
      onActivate();
    }
  }

  return /*#__PURE__*/_react.default.createElement(_ActivateOnFocus.OverlayContainer, {
    onClick: handleClick,
    onBlur: handleBlur
  }, isOverlayActive && /*#__PURE__*/_react.default.createElement(_ActivateOnFocus.FlexContainer, {
    tabIndex: 0,
    align: "center",
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_ActivateOnFocus.CardContainer, null), /*#__PURE__*/_react.default.createElement(_ActivateOnFocus.ContentContainer, null, message)), children);
}
